precompiles/erc20/tx.go

```diff
@@ -95,7 +104,18 @@ func (p Precompile) transfer(
 		return nil, err
 	}
 
-	if err := p.EmitTransferEvent(ctx, stateDB, from, to, amount); err != nil {
+	if err = p.EmitTransferEvent(ctx, stateDB, from, to, amount); err != nil {
+		return nil, err
+	}
+
+	// NOTE: if it's a direct transfer, we return here but if used through transferFrom,
+	// we need to emit the approval event with the new allowance.
+	if sender.Equals(spender) {
+		return method.Outputs.Pack(true)
+	}
+
+	newAllowance := new(big.Int).Sub(prevAllowance, amount)
+	if err = p.EmitApprovalEvent(ctx, stateDB, from, spenderAddr, newAllowance); err != nil {
 		return nil, err
 	}
```

Comment:
```suggestion
	if !sender.Equals(spender) {
		newAllowance := new(big.Int).Sub(prevAllowance, amount)
		if err = p.EmitApprovalEvent(ctx, stateDB, from, spenderAddr, newAllowance); err != nil {
			return nil, err
		}
	}
```

Reply:
@Vvaradinov I had it like this originally, but then switched the logic around to improve readability by removing the nested `if`s :v: